### PR TITLE
Members only

### PR DESCRIPTION
**Pull request by Audrey-Ella-xo**
In this project, we did the following:

1. We created the Users model and controller
2. We created a sessions controller, helpers and the login view to authenticate and authorize users that are logged in to the application.
3. We created the Post model, controllers(with index, edit and create actions) and the views for the actions.
4. We did not include a sign-up form for users because it wasn't required in the project. Instead, we added styling for the extra credit.
5. How to create users is specified in the readme file.

Thank you for your kind review. We await your speedy response

### CODE REVIEW

**Review by uche-inyama — CHANGES_REQUESTED** [review]
## Changes Required

Hi team! :wave: 

Please, referring to your comments on the pull request, these features I listed are required.

- [x] User can sign in ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/authentication/sign_in_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/authentication))

 - [x] As a signed-in user you can create new post ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/authentication/signed_in_can_post_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/authentication))

 - [x] As a signed-in user you can see post's author ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/authentication/signed_in_can_see_author_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/authentication))

 - [x] User can sign_out ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/authentication/sign_out_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/authentication))

 - [x] As a guest you can see only post's text ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/authentication/guest_can_see_only_posts_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/authentication))


Kindly update your features, then submit another request for a code review.

#### ``` Have a great day ``` :slightly_smiling_face: 

``` Uchechukwu Henry Inyama ```

**Comment by euqueme** [comment]
> * [x]  User can sign in ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/authentication/sign_in_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/authentication))

Our users can sign in please follow the instructions in our README file to create new a user and then open our project in the browser and sign in with the **sign in button**

- button 
image

- Sign in form

image

- User sign in

image

> * [x]  As a signed-in user you can create new post ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/authentication/signed_in_can_post_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/authentication))

Our sign-in users can create new posts as follows hitting the  **Create new post button**

- button

image

- Create post-form

image

- Created post

image

> * [x]  As a signed-in user you can see post's author ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/authentication/signed_in_can_see_author_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/authentication))

our signed-in users can see post's author

image

> * [x]  User can sign_out ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/authentication/sign_out_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/authentication))

Our sign-in users can sign out with the **Sign out button**

- button 

image

- result

image

> * [x]  As a guest you can see only post's text ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/authentication/guest_can_see_only_posts_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/authentication))

our guest can only see post's text, not the author

image

for the next reviewer: please :pray:  address the above :arrow_double_up:  clarifications and asks us any questions if you are having trouble installing our projects.

Sincerally @euqueme and @Audrey-Ella-xo

File: app/models/user.rb [hunk 1]
```diff
@@ -0,0 +1,31 @@
+# frozen_string_literal: true
+
+class User < ApplicationRecord
+  has_many :posts
+  attr_accessor :remember_token
+  after_create :remember
```
Comment (maelfosso) [inline app/models/user.rb]: Not needed

File: app/helpers/sessions_helper.rb [hunk 2]
```diff
@@ -0,0 +1,45 @@
+# frozen_string_literal: true
+
+module SessionsHelper
+  def log_in(user)
+    session[:user_id] = user.id
```
Comment (maelfosso) [inline app/helpers/sessions_helper.rb]: Not needed.
I suggest you avoid using `session`

File: app/models/user.rb [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+  validates :name, presence: true
+  validates :email, presence: true, uniqueness: { case_sensitive: false }
+  has_secure_password
+
+  # Returns the hash digest of the given string.
+  def self.digest(string)
+    cost = ActiveModel::SecurePassword.min_cost ? BCrypt::Engine::MIN_COST : BCrypt::Engine.cost
+    BCrypt::Password.create(string, cost: cost)
+  end
```
Comment (maelfosso) [inline app/models/user.rb]: Not needed!

File: app/models/user.rb [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+  # Returns a random token.
+  def self.new_token
+    Digest::SHA1.hexdigest SecureRandom.urlsafe_base64.to_s
+  end
+
+  # Remembers a user in the database for use in persistent sessions.
+  def remember
+    self.remember_token = User.new_token
+    update_attribute(:remember_digest, User.digest(remember_token))
```
Comment (maelfosso) [inline app/models/user.rb]: No need to call `User.digest` here. Directly use `remember_token`

File: app/controllers/sessions_controller.rb [hunk 5]
```diff
@@ -0,0 +1,28 @@
+# frozen_string_literal: true
+
+class SessionsController < ApplicationController
+  def new; end
+
+  def create
+    user = User.find_by(email: params[:session][:email].downcase)
+    if user&.authenticate(params[:session][:password])
+      log_in user
```
Comment (maelfosso) [inline app/controllers/sessions_controller.rb]: Not needed!!

Please, remove it

File: app/helpers/sessions_helper.rb [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+  end
+
+  def remember(user)
+    user.remember
+    cookies.permanent.signed[:user_id] = user.id
+    cookies.permanent[:remember_token] = user.remember_token
+  end
```
Comment (maelfosso) [inline app/helpers/sessions_helper.rb:9]: Set the `@current_user` at the end

File: app/helpers/sessions_helper.rb [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+    if (user_id = session[:user_id])
+      @current_user ||= User.find_by(id: session[:user_id])
+    elsif (user_id = cookies.signed[:user_id])
+      user = User.find_by(id: user_id)
+      if user&.authenticated?(cookies[:remember_token])
+        log_in user
+        @current_user = user
+      end
+    end
+  end
```
Comment (maelfosso) [inline app/helpers/sessions_helper.rb]: Simplify that method by only for the user with `cookies.permanent[:remember_token]`.

File: app/helpers/sessions_helper.rb [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+
+  def forget(user)
+    user.forget
+    cookies.delete(:user_id)
+    cookies.delete(:remember_token)
+  end
+
+  def log_out
+    forget(current_user)
+    session.delete(:user_id)
```
Comment (maelfosso) [inline app/helpers/sessions_helper.rb]: No session

**Review by maelfosso — CHANGES_REQUESTED** [review]
# Changes required :red_circle: 

Hello @Audrey-Ella-xo @euqueme 

Thank you for that nice job. It's working very well.
However, there some requirements that should be met to make to project approved.

2. As indicated [here](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/authentication/before_create_requirement.png), you should set a `before_create` callback method in `User` model.

3. Please, don't use `session` in this project. Instead of that use `cookies` with `remember_digest`. Read the project again for a better understanding.

4. Read the comments below

Kindly make the changes and submit a new code review request.
Thanks

**Mael FOSSO**
[Github](https://github.com) | [Twitter](https://twitter.com) | [LinkedIn](https://www.linkedin.com/in/mael-elvis-fosso-650b6346/)

File: app/controllers/sessions_controller.rb (hunk 5 above)
Comment (euqueme) [reply app/controllers/sessions_controller.rb]: removed!

File: app/models/user.rb (hunk 3 above)
Comment (euqueme) [reply app/models/user.rb]: removed!

File: app/models/user.rb (hunk 4 above)
Comment (euqueme) [reply app/models/user.rb]: modified

File: app/models/user.rb (hunk 1 above)
Comment (euqueme) [reply app/models/user.rb]: modified

File: app/helpers/sessions_helper.rb (hunk 2 above)
Comment (euqueme) [reply app/helpers/sessions_helper.rb]: removed!

File: app/helpers/sessions_helper.rb (hunk 6 above)
Comment (euqueme) [reply app/helpers/sessions_helper.rb:9]: set!

File: app/helpers/sessions_helper.rb (hunk 7 above)
Comment (euqueme) [reply app/helpers/sessions_helper.rb]: simplified

File: app/helpers/sessions_helper.rb (hunk 8 above)
Comment (euqueme) [reply app/helpers/sessions_helper.rb]: removed

File: app/models/user.rb [hunk 9]
```diff
@@ -0,0 +1,25 @@
+# frozen_string_literal: true
+
+class User < ApplicationRecord
+  has_many :posts
+  attr_accessor :remember_token
+  before_create :remember
```
Comment (Torres-ssf) [inline app/models/user.rb:6]: We need to remove this line of code. Trying to create a token for a user that is not created yet leads to an error.

**Review by Torres-ssf — APPROVED** [review]
### Project Approved :tada: 
Hi there team :hand: 
Nice job implementing the requested changes.
I'm approving this project, **but** before merging, we need to remove one line of code that's is not needed.

Good luck on the next project
Happy coding :computer: 
**Sergio Torres**

File: app/models/user.rb (hunk 9 above)
Comment (euqueme) [reply app/models/user.rb:6]: Thank you for your input but it is required in the [project](https://www.theodinproject.com/courses/ruby-on-rails/lessons/authentication) and if we don't add it somehow then the user won't have a token

image

so we changed the **remember function** so it works when creating a new user

**Pull request merged**